feat: Unit Price Items

marination opened the pull request:
Docs:
- Selling Settings: https://docs.frappe.io/erpnext/user/manual/en/selling-settings?editWiki=1&wikiPagePatch=77tg17utvv
- Buying Settings: https://docs.frappe.io/erpnext/user/manual/en/buying-settings?editWiki=1&wikiPagePatch=0hopgllcfg

Closes https://github.com/frappe/erpnext/issues/36012 (Refer for more Info)

#### Concept
- **Unit Price Items** or **Unit Price Contracts** are a way to sell/agree upon a list of items whose quantity cannot be determined at the time of the sale but Rate/Unit Price can.
- The quantity is determined once the job is fully or partially done and the agreement is to bill as per the Unit Price

<details><summary> <b>Examples:</b> </summary>

-  **Sales Order for a Consultation firm:**
<table>
<th>Item</th>
<th>Rate</th>
<th>Qty</th>

<tr>
  <td>Consulting (per hour)</td>
  <td>$60</td>
  <td>?</td>
</tr>
</table>

-  **Sales Order for a Construction Company for Road Work:**
<table>
<th>Item</th>
<th>Rate</th>
<th>Qty</th>

<tr>
  <td>Labour (per hour)</td>
  <td>$60</td>
  <td>?</td>
</tr>
<tr>
  <td>Subcontracting/Partner Costs (per hour)</td>
  <td>$40</td>
  <td>?</td>
</tr>
</table>

- The Sales Order/Quotation in both cases is agreed upon based on the unit cost/rate. 
- The ? Quantities will be determined once the job is done **E.g.** The hours put in for road work will be updated once half or all of the work is done. Invoicing is done on the basis of actual quantities.

</details>

## Setup and Working
- <img width="600" alt="Screenshot 2025-02-28 at 11 07 53 PM" src="https://github.com/user-attachments/assets/d5718687-51fb-44d0-bba7-cf383498e57c" />
- <img width="600" alt="Screenshot 2025-03-17 at 7 43 00 PM" src="https://github.com/user-attachments/assets/1ab24959-e3d0-48bd-a173-3489c2a1d990" />
- We now flag the record as `has_unit_price_items` if allowed via Settings and if it has any 0 qty items
- We don't maintain row wise flags anymore. The user is free to leave the qty as 0. Nothing more
- They will be reminded that the record contains unit price items
   <img width="700" alt="Screenshot 2025-02-28 at 11 12 59 PM" src="https://github.com/user-attachments/assets/83db109f-fbf9-4bf9-844a-535ad9083c39" />
- A row is considered a unit price item only as long as it has 0 qty
- Mapping to stock documents: Qty is mapped as is (0) until some qty is set in the Order, then it acts like a normal SO/PO row
- **Why 0 qty ?** 1 Qty might be confusing as users might assume 1 qty is being sold in actuality. Also this makes handling billing, etc. easy with the least amount of code changes

## Updating the SO/PO
Users can decide to manage their Transactions in two ways

1. Keep updating the Qty in the Order ["Sales Order", "Purchase Order"]
    - Once a portion of work is done, users can update the qty in the SO/PO -> Then create SI/PI for the updated qty
2. Keep the Qty 0 [SO, PO, RFQ, SQ, Quotation]
    - For those who might think constant updation SO is adding friction OR are not particular about tracking OR cannot update quantities in certain documents after submission, you can leave the quantities to be 0.
    - They can keep making Invoices with actual quantities against the SO/PO with 0 qty. The Billed % will always show as 0
    - **The SO/PO will have to be manually Closed**. Since there's no way to know when the Order is truly Completed.


TODO:

- [x] Extend to Purchase
- [x] Tests
- [x] Docs

> Revival of https://github.com/frappe/erpnext/pull/36024 in a hopefully cleaner way

marination committed:
feat: Unit Price Contract

marination committed:
fix: Linters

marination committed:
feat: Unit Price Items in Buying (RFQ, SQ, PO)

- chore: Extract `set_unit_price_items_note` into a util

marination committed:
test: Zero Qty in RFQ and Supplier Quotation

marination committed:
test: Purchase Order with Unit Price Items

- chore: Fix error message in accounts controller

marination committed:
test: Sales Order + fix: Mapping of Items from Quotation & SO

marination commented:
>[!CAUTION]
> Outdated comment, Ignore this.

@barredterra 
### Case:
- A Sales Order is created for 0 qty
- A DN is created for 5 qty
- Let's assume we get confirmation that 2 more units are needed and that's it
- So we set the SO Qty to **7**
- Now that the Qty is set, should this SO row behave like a normal SO row (no over-delivery allowed), or should we allow to still Deliver more and update the SO row

Currently, we maintain a hidden checkbox (to be added to each row too) `has_unit_price_items` that indicates that there are 0 qty rows
Buuut, once these rows are not 0 qty anymore does this checkbox have value? Do we want to treat the SO row as a special row where anything flies (no over-delivery limit)?

### Proposed Solution: 
If `has_unit_price_items` is set, the qty of these rows cannot be changed

### Reasoning:
- Once we move on to the DN, we don't print the Sales Order again anymore and send it to the customer
- The customer once receives the unit price sales order, but beyond that updating the qty seems to be an internal need
- Currently, allowing qty updation conveys mixed intentions which is objectively bad and leaves the correct behavior up to interpretation.

barredterra commented:
I think as long as SO qty is 0, we can deliver and/or bill any qty. As soon as SO has a qty set, be it from the start or updated at a later time, we can validate against that qty.

marination committed:
fix: Treat rows as Unit Price rows only until the qty is 0

- The unit price check should depend on the row qty being 0
- Once the row ceases to be 0, it is treated as an ordinary row
- test: PO, SO and Quotation

marination committed:
fix: Headline rendered twice on first save

- `refresh` gets triggered twice and that renders the note twice
- Remove any existing note before rendering

marination commented:
@ruthra-kumar @rohitwaghchaure Do you mind taking a look and giving this a test? Once confirmed that there are no fundamental issues I'll go ahead with docs

marination commented:
bump ⏰

ruthra-kumar commented:
Held up with other tasks. Won't be able to review this week. Will look into it next week.

marination commented:
@ruthra-kumar Any ETA possible on this? 🐱

ruthra-kumar commented:
This looks like a useful feature.

ruthra-kumar commented:
@marination 

1. Are Unit Price Sales Orders expected to have only the unit items?

2. In PR Description: 
> We don't maintain row wise flags anymore. The user is free to leave the qty as 0. Nothing more

In https://github.com/frappe/erpnext/pull/46214#issuecomment-2725553449
> Currently, we maintain a hidden checkbox (to be added to each row too)


But I don't see a row wise flag. Are you intending to add it later?

marination commented:
@ruthra-kumar Thanks for taking a look!
1. No, they can have regular items, so I'm sticking to the Unit Price **Items** terminology. 
		- An order will have a marker indicating some unit price items, and we can go easy on the 0 qty validation. 
		- Row-wise checkboxes are tedious UX wise for the user to set, and, in general, are pointless since 0 qty is the indicator
		- In use cases that I know of, there can and will be a mix of regular items too
		- This 0 qty in the order also can change once the invoices/stock transactions are finalised. That depends on the user flow.
2. I've marked the comment as dated. Sorry for the confusion, but it can be ignored. 
		- I am not planning to add a row-wise checkbox. 
		- Once the qty changes from 0 to something, the checkbox has no value, so I don't see much value in adding it.
		- I don't have a strong use case for allowing over delivery against such rows once some non-zero qty is set.

We're following this assumption https://github.com/frappe/erpnext/pull/46214#issuecomment-2725594132:
> I think as long as SO qty is 0, we can deliver and/or bill any qty. As soon as SO has a qty set, be it from the start or updated at a later time, we can validate against that qty.

marination committed:
Merge branch 'develop' into unit-price-contract-2

ruthra-kumar commented:
@marination 

> Keep updating the Qty in the Order ["Sales Order", "Purchase Order"] 

Once SI / PI is created for the whole (updated) SO / PO qty, system don't allow user to update anymore, which is normal SO / PO behavior.

What if user wants to further increase SO / PO qty?

marination commented:
@ruthra-kumar Sorry to get back a bit late. We had to confirm this case you brought up. 
- Allowing for the Updation of Items after SO/PO completion feels broader than just Unit Price Items.
- We have had requirements come up multiple times where users don't want to "Complete" an Order if it is fully invoiced/delivered, and want to keep updating the quantity and adding to it, **irrespective of Unit Price Items**.

The current solution @barredterra suggested is to:
1. Merge this PR as it is. Let's cap the scope at being able to give users the ability to not define a quantity in the Order.
2. Have a separate PR(feat) that allows to opt in for this "perpetually open order" behaviour in all cases, a sort of "Manual Completion" opt in. So this would support the case where qty is defined and could keep changing with no clear end in sight. Users will have to manually "Close" such an order.

marination committed:
Merge branch 'develop' into unit-price-contract-2

ruthra-kumar commented:
> * Merge this PR as it is. Let's cap the scope at being able to give users the ability to not define a quantity in the Order.

I agree.

> 2\. Have a separate PR(feat) that allows to opt in for this "perpetually open order" behaviour in all cases, a sort of "Manual Completion" opt in. So this would support the case where qty is defined and could keep changing with no clear end in sight. Users will have to manually "Close" such an order.

How complex will this change be?

@marination

marination commented:
> > * Merge this PR as it is. Let's cap the scope at being able to give users the ability to not define a quantity in the Order.
> 
> I agree.
> 
> > 2. Have a separate PR(feat) that allows to opt in for this "perpetually open order" behaviour in all cases, a sort of "Manual Completion" opt-in. So this would support the case where qty is defined and could keep changing with no clear end in sight. Users will have to manually "Close" such an order.
> 
> How complex will this change be?
> 
> @marination

@ruthra-kumar
1. A setting to allow this behaviour
	- If the setting is global (on a setting page): every single sales order will have to be manually closed
	- If the setting is at the form level: There's 1 extra action on each Sales Order to perform/be aware of
	- **>>** The form way makes more sense because it can be conditionally set/unset using scripts, property setters, etc., for each person's use case. If it's perpetually or conditionally set, it's up to the user
2. Tweak the status updater to mark such a Sales Order (with this setting enabled and technically "Completed") as **"Open"**. Because delivery and billing % will be 100%, no other existing status makes sense. This complements a "Closed" Sales Order well. SO/PO status shows the usual statuses and will only conditionally differ when they are "Completed"
3. Adjust form scripts to show "Update Items" if this setting is enabled. There's no API level restriction as of now.
4. I'd have to check if this behaviour breaks any existing core logic. The main concern is how such an order is interpreted by reports and other linked transactions.
5. Tweak reports. Eg. Some to-be-billed reports that rely on "status" as a query filter would deem an Open SO as something to be billed. That will have to be fixed.

marination committed:
Merge branch 'develop' into unit-price-contract-2

ruthra-kumar commented:
@marination 
For the capped scope discussed [here](https://github.com/frappe/erpnext/pull/46214#issuecomment-2834832747), PR is good to go.

I'll leave the honour of merging to you.

ruthra-kumar commented:
@marination @barredterra 

Let's hold this feature for v16 (?)
